aips/create_dip.py

```diff
@@ -0,0 +1,326 @@
+import os
+import sys
+import subprocess
+import shutil
+import uuid
+
+import metsrw
+
+from transfers import amclient
+
+THIS_DIR = os.path.abspath(os.path.dirname(__file__))
+LOGGER = logging.getLogger('create_dip')
+
+
+def setup_logger(log_file, log_level='INFO'):
+    """Configures the logger to output to console and log file"""
+    if not log_file:
+        log_file = os.path.join(THIS_DIR, 'create_dip.log')
+
+    CONFIG = {
+        'version': 1,
+        'disable_existing_loggers': True,
+        'formatters': {
+            'default': {
+                'format': '%(levelname)-8s  %(asctime)s  %(message)s',
+                'datefmt': '%Y-%m-%d %H:%M:%S',
+            },
+        },
+        'handlers': {
+            'console': {
+                'class': 'logging.StreamHandler',
+                'formatter': 'default',
+            },
+            'file': {
+                'class': 'logging.handlers.RotatingFileHandler',
+                'formatter': 'default',
+                'filename': log_file,
+                'backupCount': 2,
+                'maxBytes': 10 * 1024,
+            },
+        },
+        'loggers': {
+            'create_dip': {
+                'level': log_level,
+                'handlers': ['console', 'file'],
+            },
+        },
+    }
+
+    logging.config.dictConfig(CONFIG)
+
+
+def main(ss_url, ss_user, ss_api_key, aip_uuid, tmp_dir, output_dir):
+    LOGGER.info('Starting DIP creation from AIP: %s', aip_uuid)
+
+    if not os.path.isdir(tmp_dir):
+        LOGGER.error('%s is not a valid temporary directory', tmp_dir)
+        return
+
+    if not os.path.isdir(output_dir):
+        LOGGER.error('%s is not a valid output directory', output_dir)
+        return
+
+    # Create empty workspace directory
+    tmp_dir = os.path.join(tmp_dir, aip_uuid)
+    if os.path.exists(tmp_dir):
+        LOGGER.warning('Workspace directory already exists, overwriting')
+        shutil.rmtree(tmp_dir)
+    try:
+        os.makedirs(tmp_dir)
+    except OSError:
+        LOGGER.error('Could not create workspace directory: %s', tmp_dir)
+        return
+
+    LOGGER.info('Downloading AIP from Storage Service')
+
+    am_client = amclient.AMClient(
+        aip_uuid=aip_uuid,
+        ss_url=ss_url,
+        ss_user_name=ss_user,
+        ss_api_key=ss_api_key,
+        directory=tmp_dir)
+
+    aip_file = am_client.download_aip()
+
+    if not aip_file:
+        LOGGER.error('Unable to download AIP')
+        return
+
+    LOGGER.info('Extracting AIP')
+    aip_dir = extract_aip(aip_file, aip_uuid, tmp_dir)
+
+    if not aip_dir:
+        return
+
+    LOGGER.info('Creating DIP')
+    dip_dir = create_dip(aip_dir, aip_uuid, output_dir)
+
+    if not dip_dir:
+        LOGGER.error('Unable to create DIP')
+        return
+
+    LOGGER.info('DIP created in: %s', dip_dir)
+
+
+def extract_aip(aip_file, aip_uuid, tmp_dir):
+    """Extract a downloaded AIP to a folder."""
```

Comment:
You could explain here something about the returned value like in [this example](https://github.com/artefactual-labs/mets-reader-writer/blob/21438f284fcb97a78718b5feee16b0b947503450/metsrw/fsentry.py#L347-L362).

Reply:
On a side note, I think sometimes it's more elegant to let an exception bubble up. That would be the Pythonic way in Python I believe to communicate the error to the caller. The caller can control the exception inside a try/except and do the logging for you so your function becomes even smaller. In some cases you may even want to raise your own exception, e.g. `if not extracted_entry: raise Exception('Cannot find extracted AIP by UUID')`.